File: dotnet/picocontainer/src/Extras/DefaultLifecyclePicoAdapter.cs
/*****************************************************************************
 * Copyright (C) PicoContainer Organization. All rights reserved.            *
 * ------------------------------------------------------------------------- *
 * The software in this package is published under the terms of the BSD      *
 * style license a copy of which has been included with this distribution in *
 * the license.txt file.                                                     *
 *                                                                           *
 * Idea by Rachel Davies, Original code by Aslak Hellesoy and Paul Hammant   *
 * C# port by Maarten Grootendorst                                           *
 *****************************************************************************/

using System;
using PicoContainer.Lifecycle;

namespace PicoContainer.Extras
{
	public class DefaultLifecyclePicoAdapter : LifecyclePicoAdapter
	{
		virtual public bool Started
		{
			get
			{
				return started;
			}
			
		}
		virtual public bool Stopped
		{
			get
			{
				return !started;
			}
			
		}
    
		virtual public bool Disposed
		{
			get
			{
				return disposed;
			}
			
		}
		virtual public PicoContainer PicoContainer
		{
			get
			{
				return picoContainer;
			}
			
		}
		
		private Startable StartableAggregatedComponent;
		private Stoppable StoppableAggregatedComponent;
		private IDisposable disposableAggregatedComponent;
		private bool started;
		private bool disposed;

		private PicoContainer picoContainer;
		
		public DefaultLifecyclePicoAdapter(PicoContainer picoContainer)
		{
			this.picoContainer = picoContainer;
		}
		
		private void  InitializeIfNotInitialized()
		{
			if (StartableAggregatedComponent == null)
			{
			  StartableAggregatedComponent = picoContainer.GetComponentMulticaster(true, false) as Startable ;
			}
			if (StoppableAggregatedComponent == null)
			{
        StoppableAggregatedComponent = picoContainer.GetComponentMulticaster(false, false) as Stoppable;
			}
			if (disposableAggregatedComponent == null)
			{
					disposableAggregatedComponent = picoContainer.GetComponentMulticaster(false, false) as IDisposable;
			}
		}
		
		public virtual void  Start()
		{
			CheckDisposed();
			InitializeIfNotInitialized();
			if (Started)
			{
				throw new System.SystemException("Already Started.");
			}
			started = true;
			if (StartableAggregatedComponent != null)
			{
				StartableAggregatedComponent.Start();
			}
		}
		
		public virtual void  Stop()
		{
			CheckDisposed();
			InitializeIfNotInitialized();
			if (started == false)
			{
				throw new System.SystemException("Already Stopped (or maybe never Started).");
			}
			started = false;
			if (StoppableAggregatedComponent != null)
			{
				StoppableAggregatedComponent.Stop();
			}
		}
		
		public virtual void  Dispose()
		{
			CheckDisposed();
			InitializeIfNotInitialized();
			disposed = true;
			if (disposableAggregatedComponent != null)
			{
				disposableAggregatedComponent.Dispose();
			}
		}
		
		private void  CheckDisposed()
		{
			if (disposed)
			{
				throw new System.SystemException("Components already disposed of");
			}
		}
	}
}

File: java/picocontainer/src/java/org/picocontainer/extras/DefaultLifecyclePicoAdapter.java
/*****************************************************************************
 * Copyright (C) PicoContainer Organization. All rights reserved.            *
 * ------------------------------------------------------------------------- *
 * The software in this package is published under the terms of the BSD      *
 * style license a copy of which has been included with this distribution in *
 * the license.html file.                                                    *
 *                                                                           *
 * Idea by Rachel Davies, Original code by Aslak Hellesoy and Paul Hammant   *
 *****************************************************************************/

package org.picocontainer.extras;

import org.picocontainer.PicoContainer;
import org.picocontainer.PicoException;
import org.picocontainer.PicoIntrospectionException;
import org.picocontainer.defaults.AssignabilityRegistrationException;
import org.picocontainer.defaults.DefaultComponentMulticasterPicoAdapter;
import org.picocontainer.defaults.NotConcreteRegistrationException;
import org.picocontainer.lifecycle.Disposable;
import org.picocontainer.lifecycle.LifecyclePicoAdapter;
import org.picocontainer.lifecycle.Startable;
import org.picocontainer.lifecycle.Stoppable;

/**
 * @author Paul Hammant
 * @author Ward Cunningham
 * @version $Revision$
 */
public class DefaultLifecyclePicoAdapter implements LifecyclePicoAdapter {

    private Startable startableAggregatedComponent;
    private Stoppable stoppableAggregatedComponent;
    private Disposable disposableAggregatedComponent;
    private boolean started;
    private boolean disposed;
    private final PicoContainer picoContainer;
    private final ComponentMulticasterPicoAdapter multicasterAdapter;

    public DefaultLifecyclePicoAdapter(PicoContainer picoContainer, ComponentMulticasterPicoAdapter multicasterAdapter) {
         this.picoContainer = picoContainer;
         this.multicasterAdapter = multicasterAdapter;
    }

    public DefaultLifecyclePicoAdapter(PicoContainer picoContainer) {
        this.picoContainer = picoContainer;
        if(picoContainer instanceof ComponentMulticasterPicoAdapter) {
            multicasterAdapter = (ComponentMulticasterPicoAdapter)picoContainer;
        } else {
            multicasterAdapter = new DefaultComponentMulticasterPicoAdapter(picoContainer);
        }
    }

    public boolean isStarted() {
        return started;
    }

    public boolean isStopped() {
        return !started;
    }

    public boolean isDisposed() {
        return disposed;
    }

    public PicoContainer getPicoContainer() {
        return picoContainer;
    }

    private void initializeIfNotInitialized() throws PicoException, PicoIntrospectionException, AssignabilityRegistrationException, NotConcreteRegistrationException {
        if (startableAggregatedComponent == null) {
            try {
                startableAggregatedComponent = (Startable) multicasterAdapter.getComponentMulticaster(true, false);
            } catch (ClassCastException e) {
            }
        }
        if (stoppableAggregatedComponent == null) {
            try {
                stoppableAggregatedComponent = (Stoppable) multicasterAdapter.getComponentMulticaster(false, false);
            } catch (ClassCastException e) {
            }
        }
        if (disposableAggregatedComponent == null) {
            try {
                Object o = multicasterAdapter.getComponentMulticaster(false, false);
                disposableAggregatedComponent = (Disposable) o;
            } catch (ClassCastException e) {
            }
        }

    }

    public void start() {
        checkDisposed();
        initializeIfNotInitialized();
        if (started) {
            throw new IllegalStateException("Already started.");
        }
        started = true;
        if (startableAggregatedComponent != null) {
            startableAggregatedComponent.start();
        }
    }

    public void stop() {
        checkDisposed();
        initializeIfNotInitialized();
        if (started == false) {
            throw new IllegalStateException("Already stopped (or maybe never started).");
        }
        started = false;
        if (stoppableAggregatedComponent != null) {
            stoppableAggregatedComponent.stop();
        }
    }

    public void dispose() {
        checkDisposed();
        initializeIfNotInitialized();
        disposed = true;
        if (disposableAggregatedComponent != null) {
            disposableAggregatedComponent.dispose();
        }
    }

    private void checkDisposed() {
        if (disposed) {
            throw new IllegalStateException("Components already disposed of");
        }
    }

}

